[UI-side compositng] Use ScrollbarsController base class to avoid NSScrollerImp usage

nmoucht opened the pull request:
#### 0b438d0d3534b730070191525a8a0508ccc633d2
<pre>
[UI-side compositng] Use ScrollbarsController base class to avoid NSScrollerImp usage
<a href="https://bugs.webkit.org/show_bug.cgi?id=253684">https://bugs.webkit.org/show_bug.cgi?id=253684</a>
rdar://106531543

Reviewed by Simon Fraser.

Have FrameView and RenderLayerScrollable area hold a base class ScrollbarsController
to avoid having NSScrollerImps in the WebProcess.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::usesUISideCompositing const):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::didAddScrollbar):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarsController const):
(WebCore::ScrollableArea::didAddScrollbar):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h: Added.
* Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.mm: Added.
(WebCore::ScrollbarsController::createProxy):
(WebCore::ScrollbarsControllerProxyMac::ScrollbarsControllerProxyMac):
(WebCore::ScrollbarsControllerProxyMac::~ScrollbarsControllerProxyMac):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::usesUISideCompositing const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: <a href="https://commits.webkit.org/262176@main">https://commits.webkit.org/262176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e69c8519bd157c577b901cb4371ab396a8511b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/436 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/455 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/423 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/377 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/422 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/497 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/623 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/408 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/455 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/449 "3 api tests failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/397 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/457 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/367 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/413 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/396 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/406 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/408 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/404 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/84 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -68,15 +71,20 @@ ScrollAnimator& ScrollableArea::scrollAnimator() const
-        } else
-            m_scrollbarsController = ScrollbarsController::create(const_cast<ScrollableArea&>(*this));
+        } else {
+            if (page->chrome().client().isUsingUISideCompositing()) {
```
If page can be null, we need to null check page here.

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,53 @@
+public:
+    explicit ScrollbarsControllerProxyMac(ScrollableArea&, ScrollingCoordinator*);
+    ~ScrollbarsControllerProxyMac();
+    ScrollingCoordinator* m_coordinator;
```
This should be a WeakPtr unless there is an obvious guarantee of the same lifetime.

nmoucht replied:
Fixed.

rr-codes commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,52 @@
+public:
+    explicit ScrollbarsControllerProxyMac(ScrollableArea&, ScrollingCoordinator*);
+    ~ScrollbarsControllerProxyMac();
+    WeakPtr<ScrollingCoordinator> m_coordinator;
```
should this member be public?

cdumez replied:
I agree this should be definitely private.

cdumez commented on Source/WebCore/page/scrolling/ScrollingCoordinator.h:
```diff
@@ -60,7 +60,7 @@ class ViewportConstraints;
 using PlatformDisplayID = uint32_t;
 
-class ScrollingCoordinator : public ThreadSafeRefCounted<ScrollingCoordinator> {
+class ScrollingCoordinator : public ThreadSafeRefCounted<ScrollingCoordinator>, public CanMakeWeakPtr<ScrollingCoordinator> {
```
`ThreadSafeRefCounted` + `CanMakeWeakPtr` should be `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr`

cdumez commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -68,15 +71,20 @@ ScrollAnimator& ScrollableArea::scrollAnimator() const
+        } else {
+            if (page && page->chrome().client().isUsingUISideCompositing()) {
+                m_scrollbarsController = ScrollbarsController::createProxy(const_cast<ScrollableArea&>(*this), page->scrollingCoordinator());
+
```
Extra line here.

cdumez commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -195,7 +196,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea>, public CanMakeChec
     ScrollAnimator* existingScrollAnimator() const { return m_scrollAnimator.get(); }
 
-    WEBCORE_EXPORT ScrollbarsController& scrollbarsController() const;
+    WEBCORE_EXPORT ScrollbarsController& scrollbarsController(Page* page = nullptr) const;
```
`page` can/should be omitted

cdumez commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -165,7 +166,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea>, public CanMakeChec
     WEBCORE_EXPORT bool scrollbarsCanBeActive() const;
 
-    WEBCORE_EXPORT virtual void didAddScrollbar(Scrollbar*, ScrollbarOrientation);
+    WEBCORE_EXPORT virtual void didAddScrollbar(Scrollbar*, ScrollbarOrientation, Page* page = nullptr);
```
`page` can/should be omitted

cdumez commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,52 @@
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(ScrollbarsControllerProxyMac);
+public:
+    explicit ScrollbarsControllerProxyMac(ScrollableArea&, ScrollingCoordinator*);
```
There are 2 mandatory parameters, is there a benefit to marking this explicit?

cdumez commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,52 @@
+
+class WheelEventTestMonitor;
+
+class ScrollbarsControllerProxyMac : public ScrollbarsController {
```
Can this class be marked as final?

cdumez commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,52 @@
+class WheelEventTestMonitor;
+
+class ScrollbarsControllerProxyMac : public ScrollbarsController {
+    WTF_MAKE_FAST_ALLOCATED;
```
Not needed, the base class already has this.

cdumez commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,52 @@
+
+class ScrollbarsControllerProxyMac : public ScrollbarsController {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(ScrollbarsControllerProxyMac);
```
ditto.

cdumez commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.mm:
```diff
@@ -0,0 +1,64 @@
+{
+}
+
+ScrollbarsControllerProxyMac::~ScrollbarsControllerProxyMac()
```
= default; ?

smfr commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -68,15 +71,19 @@ ScrollAnimator& ScrollableArea::scrollAnimator() const
 }
 
-ScrollbarsController& ScrollableArea::scrollbarsController() const
+ScrollbarsController& ScrollableArea::scrollbarsController(Page* page) const
```
This is a layering violation; we're down in `/platform/` here and can't know about `Page`.

smfr commented on Source/WebCore/page/ChromeClient.cpp:
```diff
@@ -30,6 +30,8 @@
 #include "GraphicsContextGL.h"
 #endif
 
+#include "ScrollbarsController.h"
```
This should go above the #ifdeffed includes.

smfr commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -389,7 +389,7 @@ class ChromeClient {
     // regardless of the settings.
     virtual bool allowsAcceleratedCompositing() const { return true; }
-
+        
```
Whitespace.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -5803,7 +5804,13 @@ void LocalFrameView::setScrollingPerformanceTestingEnabled(bool scrollingPerform
     Page* page = m_frame->page();
+    if (page) {
+        if (auto scrollbarController = page->chrome().client().createScrollbarsController(*page, *this))
+            setScrollbarsController(scrollbarController);
```
`WTFMove(scrollbarController)`

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinator.h:
```diff
@@ -61,7 +61,7 @@ struct KeyboardScroll;
 using PlatformDisplayID = uint32_t;
 
-class ScrollingCoordinator : public ThreadSafeRefCounted<ScrollingCoordinator> {
+class ScrollingCoordinator : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<ScrollingCoordinator> {
```
It's not clear why this is necessary

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -51,6 +51,7 @@ class ScrollAnimator;
 class ScrollbarsController;
 class GraphicsLayer;
 class TiledBacking;
+class Page;
```
Not needed (and a layering violation).

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -196,6 +197,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea>, public CanMakeChec
     ScrollAnimator* existingScrollAnimator() const { return m_scrollAnimator.get(); }
 
     WEBCORE_EXPORT ScrollbarsController& scrollbarsController() const;
+    void setScrollbarsController(std::unique_ptr<ScrollbarsController>&);
```
Argument should be a `std::unique_ptr<ScrollbarsController>&&`

smfr commented on Source/WebCore/platform/ScrollbarsController.cpp:
```diff
@@ -37,6 +37,13 @@ std::unique_ptr<ScrollbarsController> ScrollbarsController::create(ScrollableAre
+#if !PLATFORM(MAC)
+std::unique_ptr<ScrollbarsController> ScrollbarsController::createProxy(ScrollableArea& scrollableArea, ScrollingCoordinator*)
+{
+    return makeUnique<ScrollbarsController>(scrollableArea);
```
It's weird that this just creates another `ScrollbarsController`?

smfr commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -34,12 +34,14 @@ namespace WebCore {
 public:
-    static std::unique_ptr<ScrollbarsController> create(ScrollableArea&);
+    WEBCORE_EXPORT static std::unique_ptr<ScrollbarsController> create(ScrollableArea&);
+    WEBCORE_EXPORT static std::unique_ptr<ScrollbarsController> createProxy(ScrollableArea&, ScrollingCoordinator*);
```
The concept of "proxy" objects is generally used only between processes, which is not something that WebCore should know about.

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h:
```diff
@@ -0,0 +1,51 @@
+
+class WheelEventTestMonitor;
+
+class ScrollbarsControllerProxyMac final : public ScrollbarsController {
```
This "proxy" class should not live in WebCore; if it's going to be that UI-process representation of something, it should live up in WebKit.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1034,7 +1041,7 @@ bool WebChromeClient::layerTreeStateIsFrozen() const
 #if ENABLE(ASYNC_SCROLLING)
 
-RefPtr<ScrollingCoordinator> WebChromeClient::createScrollingCoordinator(Page& page) const
+RefPtr<WebCore::ScrollingCoordinator> WebChromeClient::createScrollingCoordinator(Page& page) const
```
This file has `using namespace WebCore`. You don't need the `WebCore::` namespace here or below.

smfr reviewed (changes requested):
I'm not really clear on where this patch is going; I'm not convinced that ScrollbarControllers should be IPCing back and forth, since scrolling tree commits could already send the state of scrollbars from web to UI process.

nmoucht committed:
[UI-side compositng] Use ScrollbarsController base class to avoid NSScrollerImp usage
https://bugs.webkit.org/show_bug.cgi?id=253684
rdar://106531543

Reviewed by Simon Fraser.

Have FrameView and RenderLayerScrollable area hold a base class ScrollbarsController
to avoid having NSScrollerImps in the WebProcess.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::usesUISideCompositing const):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::didAddScrollbar):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarsController const):
(WebCore::ScrollableArea::didAddScrollbar):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.h: Added.
* Source/WebCore/platform/mac/ScrollbarsControllerProxyMac.mm: Added.
(WebCore::ScrollbarsController::createProxy):
(WebCore::ScrollbarsControllerProxyMac::ScrollbarsControllerProxyMac):
(WebCore::ScrollbarsControllerProxyMac::~ScrollbarsControllerProxyMac):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::usesUISideCompositing const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: https://commits.webkit.org/262176@main
